### PR TITLE
Refactor fun setNodeStatus

### PR DESCRIPTION
**Pull request by zhengguoyong**
setNodeStatus is so long, it can split five steps:
1: set the address for the node;
2: set the status info for the node;
3: set the ready condition for the node;
4: set the out of disk condition for the node;
5: record event if the node schdule changed;
i just put each step into a function and other tiny changes to make it clean;

### CODE REVIEW

**Comment by yujuhong** [comment]
@zhengguoyong I like the direction, but #14542 might touch the same code. I will review after #14542 lands. Thanks!

**Comment by zhengguoyong** [comment]
ok, i will rebase it first, thanks.  
@yujuhong

**Comment by zhengguoyong** [comment]
PR(#14542) is merged now, please help review this PR.
thanks.   :)
@yujuhong

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2593,12 +2589,28 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 		// TODO: kube-proxy might be different version from kubelet in the future
 		node.Status.NodeInfo.KubeProxyVersion = version.Get().String()
 	}
+	return
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Don't need to add an explicit `return`.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2593,12 +2589,28 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+}
 
+// Set daemonEndpoints for the node.
+func (kl *Kubelet) setNodeStatusDaemonEndpoints(node *api.Node) {
 	node.Status.DaemonEndpoints = *kl.daemonEndpoints
+	return
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Don't need to add an explicit `return`.

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -2593,12 +2589,28 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+}
+
+// Set status for the node.
+func (kl *Kubelet) setNodeStatusInfo(node *api.Node) {
+	kl.setNodeStatusMachineInfo(node)
+	kl.setNodeStatusVersionInfo(node)
+	kl.setNodeStatusDaemonEndpoints(node)
+	return
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Don't need to add an explicit `return`.

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -2617,33 +2629,41 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
 	}
-	if !updated || oldNodeReadyConditionStatus != newNodeReadyCondition.Status {
+	if needRecordEvent {
 		if newNodeReadyCondition.Status == api.ConditionTrue {
 			kl.recordNodeStatusEvent(kubecontainer.NodeReady)
 		} else {
 			kl.recordNodeStatusEvent(kubecontainer.NodeNotReady)
 		}
 	}
+	return
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Don't need to add an explicit `return`.

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -2617,33 +2629,41 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 		}
 	}
 
-	updated := false
+	needUpdated := true
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: I don't think the variable name is clear. Why don't we change to to `readyConditionUpdated := false`. I think the original default (false) is also more intuitive.

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -2617,33 +2629,41 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+	needRecordEvent := true
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: We only need to record an event when there is a transition. I prefer setting the default to false, and flip the variable after line 2641. Also, rename the variable to `needToRecordEvent` or `shouldRecordEvent`.

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -2675,28 +2695,33 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+		nodeOODCondition.Reason = "KubeletHasSufficientDisk"
+		nodeOODCondition.Message = "kubelet has sufficient disk space available"
+		nodeOODCondition.LastTransitionTime = currentTime
+		kl.recordNodeStatusEvent("NodeHasSufficientDisk")
 	}
 
 	if newOODCondition {
 		node.Status.Conditions = append(node.Status.Conditions, *nodeOODCondition)
 	}
+	return
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: same as above, remove the `return`.

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -2705,6 +2730,21 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 		}
 		oldNodeUnschedulable = node.Spec.Unschedulable
 	}
+	return
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: no need to return

File: pkg/kubelet/kubelet.go [hunk 9]
```diff
@@ -2675,28 +2695,33 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
 	}
 
 	if newOODCondition {
 		node.Status.Conditions = append(node.Status.Conditions, *nodeOODCondition)
 	}
+	return
+}
 
+// Maintains Node.Spec.Unschedulable value from previous run of tryUpdateNodeStatus()
+var oldNodeUnschedulable bool
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Hmm...I don't like the global variable here. I prefer the old logic.  Also, why factoring out this as a function when you don't do that for the node ready status event.

File: pkg/kubelet/kubelet.go [hunk 10]
```diff
@@ -2675,28 +2695,33 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
...
+
+// record if node schedulable change.
+func (kl *Kubelet) recordNodeschduleEvent(node *api.Node) {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: s/recordNodeschduleEvent/recordNodeSchdulableEvent

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (zhengguoyong) [inline pkg/kubelet/kubelet.go]: Done. :)

File: pkg/kubelet/kubelet.go (hunk 5 above)
Comment (zhengguoyong) [inline pkg/kubelet/kubelet.go]: Done.  :)

File: pkg/kubelet/kubelet.go (hunk 6 above)
Comment (zhengguoyong) [inline pkg/kubelet/kubelet.go]: Done. :)

File: pkg/kubelet/kubelet.go (hunk 9 above)
Comment (zhengguoyong) [inline pkg/kubelet/kubelet.go]: this global variable is exist before, i just move here to let it closer to where it used.
:)

File: pkg/kubelet/kubelet.go (hunk 10 above)
Comment (zhengguoyong) [inline pkg/kubelet/kubelet.go]: Done. :)

**Comment by zhengguoyong** [comment]
Hi @yujuhong ,
thanks very much for your review first.
i have changed basical on your views.
please help review again,thanks.  :)

**Comment by zhengguoyong** [comment]
@k8s-bot  test this please

**Comment by zhengguoyong** [comment]
@k8s-bot test this please

**Comment by zhengguoyong** [comment]
@k8s-bot test this please

File: pkg/kubelet/kubelet.go (hunk 9 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: :-(

**Comment by yujuhong** [comment]
@zhengguoyong, LGTM, could you rebase? Thanks!

**Comment by zhengguoyong** [comment]
I did the rebase,
Please reivew again,
Thanks.  
@yujuhong

**Pull request merged**